flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteTierProducerAgent.java

```diff
@@ -0,0 +1,110 @@
+
+    private final int numBuffersPerSegment;
+
+    private final RemoteCacheManager cacheDataManager;
+
+    private final int[] currentSubpartitionWriteBuffers;
+
+    RemoteTierProducerAgent(
+            TieredStoragePartitionId partitionId,
+            int numSubpartitions,
+            int numBytesPerSegment,
+            int bufferSizeBytes,
+            boolean isBroadcastOnly,
+            PartitionFileWriter partitionFileWriter,
+            TieredStorageMemoryManager memoryManager,
+            TieredStorageResourceRegistry resourceRegistry) {
+        checkArgument(
+                numBytesPerSegment >= bufferSizeBytes,
+                "One segment should contain at least one buffer.");
+
+        this.numSubpartitions = numSubpartitions;
+        this.numBuffersPerSegment = numBytesPerSegment / bufferSizeBytes;
+        this.cacheDataManager =
+                new RemoteCacheManager(
+                        partitionId,
+                        isBroadcastOnly ? 1 : numSubpartitions,
+                        memoryManager,
+                        partitionFileWriter);
+        this.currentSubpartitionWriteBuffers = new int[numSubpartitions];
+        Arrays.fill(currentSubpartitionWriteBuffers, 0);
+        resourceRegistry.registerResource(partitionId, this::releaseAllResources);
+    }
+
+    @Override
+    public boolean tryStartNewSegment(TieredStorageSubpartitionId subpartitionId, int segmentId) {
+        cacheDataManager.startSegment(subpartitionId.getSubpartitionId(), segmentId);
+        // The remote storage tier should always be able to start a new segment.
+        return true;
+    }
+
+    @Override
+    public boolean tryWrite(TieredStorageSubpartitionId subpartitionId, Buffer buffer) {
+        int subpartitionIndex = subpartitionId.getSubpartitionId();
+        if (currentSubpartitionWriteBuffers[subpartitionIndex] != 0
+                && currentSubpartitionWriteBuffers[subpartitionIndex] + 1 > numBuffersPerSegment) {
+            cacheDataManager.finishSegment(subpartitionIndex);
+            currentSubpartitionWriteBuffers[subpartitionIndex] = 0;
+            return false;
+        }
+        currentSubpartitionWriteBuffers[subpartitionIndex]++;
+        emitBuffer(buffer, subpartitionIndex);
+        return true;
+    }
+
+    @Override
+    public void close() {
+        for (int subpartitionId = 0; subpartitionId < numSubpartitions; subpartitionId++) {
+            cacheDataManager.finishSegment(subpartitionId);
+        }
+        cacheDataManager.close();
+    }
+
+    // ------------------------------------------------------------------------
+    //  Internal Methods
+    // ------------------------------------------------------------------------
+
+    private void releaseAllResources() {
+        cacheDataManager.release();
+    }
+
+    private void emitBuffer(Buffer finishedBuffer, int subpartitionId) {
+        cacheDataManager.appendBuffer(finishedBuffer, subpartitionId);
+    }
```

Comment:
1. The name `emit` is confusing.
2. Why do we need this method when its body is only one line of code?

Reply:
Fixed.